src/lib.rs

```diff
@@ -1,9 +1,85 @@
+#![allow(rustdoc::private_intra_doc_links)]
+/*!
+
+Easy to use profiler scopes for [wgpu](https://github.com/gfx-rs/wgpu) using timer queries.
+
+wgpu_profile manages all the necessary [wgpu::QuerySet] and [wgpu::Buffer] behind the scenes
+and allows you to create to create timer scopes with minimal overhead!
+
+# How to use
+
+```
+use wgpu_profiler::*;
+
+# async fn wgpu_init() -> (wgpu::Instance, wgpu::Adapter, wgpu::Device, wgpu::Queue) {
+    # let instance = wgpu::Instance::new(wgpu::Backends::all());
+    # let adapter = instance.request_adapter(&wgpu::RequestAdapterOptions::default()).await.unwrap();
+    # let (device, queue) = adapter
+    #     .request_device(
+    #         &wgpu::DeviceDescriptor {
+    #             features: wgpu::Features::TIMESTAMP_QUERY,
+    #             ..Default::default()
+    #         },
+    #         None,
+    #     )
+    #     .await
+    #     .unwrap();
+    # (instance, adapter, device, queue)
+# }
+# let (instance, adapter, device, queue) = futures_lite::future::block_on(wgpu_init());
+// ...
+
+let mut profiler = GpuProfiler::new(4, queue.get_timestamp_period(), device.features());
+
+// ...
+
+# let mut encoder = device.create_command_encoder(&wgpu::CommandEncoderDescriptor::default());
+// Using scopes is easiest with the macro:
+wgpu_profiler!("name of your scope", &mut profiler, &mut encoder, &device, {
+  // wgpu commands go here
+});
+
+// Wgpu-profiler needs to insert buffer copy commands.
+profiler.resolve_queries(&mut encoder);
+# drop(encoder);
+
+// ...
+
+// And finally, to end a profiling frame, call `end_frame`.
+// This does a few checks and will let you know of something is off!
+profiler.end_frame().unwrap();
+
+// Retrieving the oldest available frame and writing it out to a chrome trace file.
+if let Some(profiling_data) = profiler.process_finished_frame() {
+    # let button_pressed = false;
+    // You usually want to write to disk only under some condition, e.g. press of a key.
+    if button_pressed {
+        wgpu_profiler::chrometrace::write_chrometrace(
+            std::path::Path::new("mytrace.json"), &profiling_data);
+    }
+}
+```
+Check also the [Example](https://github.com/Wumpf/wgpu-profiler/blob/main/examples/demo.rs) where everything can be seen in action.
+
+# Internals
+
+For every frame that hasn't completely finished processing yet
+(i.e. hasn't returned results via [process_finished_frame](GpuProfiler::process_finished_frame))
+we keep a [PendingFrame] around.
```

Comment:
Since they are private items, `PendingFrame` and `QueryPool` links look like this:
![image](https://user-images.githubusercontent.com/2002109/190880220-766ddd9d-9a6d-41e0-837a-2b1156a84568.png)

Reply:
Unless though they are compiled with `cargo doc --document-private-items`. But I guess that's too uncommon :/